go-metrics: pull library and introduce ResettingTimer and InfluxDB reporter.

nonsense opened the pull request:
This PR introduces the following changes:
1. updates the `rcrowley/go-metrics` to its latest version
2. fork `github.com/rcrowley/go-metrics` to `github.com/ethersphere/go-metrics`, in order to:
3. add a `ResettingTimer` metric type which calculates `mean`, `50%ile`, `95%ile`, `99%ile`, `min`, and `max` for timers per flush interval - timers are reset on every `Snapshot()` call.
4. vendor `github.com/ethersphere/go-metrics-influxdb`, which includes an InfluxDB reporter - to be used to flush metrics registry to an InfluxDB backend every N seconds.

These changes are necessary because we want to be able to measure latencies in `swarm` (and probably at some point in `geth`?) and forward them to a time-series backend (InfluxDB) and visualise them later with Grafana. The existing `Timer` type is not suitable for this, see http://taint.org/2014/01/16/145944a.html or https://groups.google.com/d/msg/mechanical-sympathy/I4JfZQ1GYi8/Bf1ftrjpuKMJ .

TODO:
- [x] add tests for `ResettingTimer`
- [ ] use the up-to-date InfluxDB client at https://github.com/influxdata/influxdb/tree/master/client/v2 instead of https://github.com/influxdata/influxdb/tree/master/client which is deprecated.
- [x] move logic which calculates `mean`, `50%ile`, `95%ile`, `99%ile`, `min`, and `max` away from the reporter into the `ResettingTimer`.

lmars commented on consensus/ethash/ethash.go:
```diff
@@ -35,7 +35,7 @@ import (
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/rpc"
-	metrics "github.com/rcrowley/go-metrics"
+	metrics "github.com/nonsense/go-metrics"
```
Should we fork to `ethereum` or `ethersphere` instead so the team can manage this dependency? I've just hit this issue in the past where a user's fork needs to be forked again to make changes.

nonsense replied:
Good point. I'm happy with any repo. `ethersphere` or `ethereum` would both work. If we don't want to keep `go-metrics` and `go-metrics-influxdb` on `ethereum`, I would be happy to fork them on `ethersphere` so that at least there are more people who have access to it.

nonsense replied:
Actually my first thought was to vendor it like the `log` package which you pointed out, but there is already a `metrics` package which does other things, and having the `go-metrics` under the `metrics` folder seemed weird. But maybe we can vendor all the metrics related libraries under `metrics/lib` in go-ethereum?

karalabe replied:
Why do we need to fork go-metrics? Can't we just add one more metric of our own (`ResettingTimer`)?

nonsense replied:
We would also have to create our own `Registry` (see https://github.com/ethereum/go-ethereum/pull/15910/files#diff-69f4fcff395026c7300bbaa7c953e251R199), when the `StandardRegistry` is good for our purposes.

nonsense replied:
In my opinion `metrics` library is something that doesn't change often, so I'd rather keep all related code in one library, rather than patch additional functionality with external structs. Because of this I found forking to be the easiest. I'm happy to hear other opinions on this though.

nonsense replied:
@lmars changed fork location to `ethersphere`.

nonsense commented on vendor/github.com/ethersphere/go-metrics/resetting_timer.go:
```diff
@@ -0,0 +1,237 @@
+package metrics
```
This is the main change in this PR. Since `govendor` does not include tests, you can find them at https://github.com/ethersphere/go-metrics/blob/master/resetting_timer_test.go

nonsense commented:
@lmars @karalabe I rebased this on `master` so that it is up-to-date. AppVeyor seems to be crashing with unrelated code:
```
[00:21:22] fatal error: out of memory
[00:21:22] FAIL	github.com/ethereum/go-ethereum/tests	209.548s
```

Is there anything else you expect from this PR? Are you happy with it in general, or shall I fix something before we merge?

lmars commented:
LGTM

nonsense commented:
@karalabe as discussed I pulled the `go-metrics` repo in `go-ethereum`, rather than vendoring it.

The changes introduced are as follows:
1. `go-metrics` had `UseNilMetrics` global bool, which was identical to `go-ethereum`'s `Enabled`, so I got rid of `UseNilMetrics` and used `Enabled` instead.
2. `go-ethereum` code used a custom constructor for measurements, which was in contradiction with the `go-metrics` API, so I changed all NewCounter, etc. to NewRegisteredCounter passing the default registry.
3. Introduced package `exp` in order to solve a circular dependency and init the `metrics/exp` package.
4. Pulled in the `influxdb` reporter, as it is just one file, and doesn't make sense to maintain a fork for it either.

cc @lmars

karalabe commented on exp/exp.go:
```diff
@@ -0,0 +1,10 @@
+
+func init() {
+	e.Exp(metrics.DefaultRegistry)
+}
```
I don't think you wanted to commit this file :)

karalabe replied:
Or I don't get it why this is outside in its own package vs underneath metrics?

nonsense replied:
Currently we have https://github.com/ethereum/go-ethereum/blob/master/metrics/metrics.go#L48 . Since we are merging go-ethereum metrics and go-metrics in one package, we can't have this initialisation in the new joint `metrics` as it results in circular dependency.

nonsense replied:
Then again I don't think we want a separate package just for this, so it can go anywhere else but in the `metrics` package.

karalabe replied:
Well, since it's our custom package now, we can just do this call within `metrics/exp/exp.go` as an init. Would that be bad?

nonsense replied:
Sure, I will move it to `metrics/exp/exp.go`. What I am not very happy about is that we are starting an HTTP server in this file, which is kind of hidden away, and we would always be initialising the /debug/metrics endpoint with the default registry.

I can imagine that at some point we might want to have more than one registry (default registry, swarm registry, etc.), but I guess for now every binary can just use the default registry for simplicity.

So for now `metrics/exp/exp.go` should be fine to init this server, but we might want to refactor in the future.

nonsense replied:
This endpoint is exposed as part of the pprof http server (see https://github.com/ethereum/go-ethereum/commit/9e9bfc4e264b0bc16bdfe893e9cb3cfad096c48c), so I decided to set it up next to the pprof http server.

Moving it to `metrics/exp/exp.go` is not a solution, because no package is including `exp` so the `init` would never be called. We need another package to import the `exp` package, so `internal/debug` seems to be the correct place.

@karalabe I hope that makes sense.
